In [72]:
pip install pandas xlrd ipython-sql sqlalchemy openpyxl

Note: you may need to restart the kernel to use updated packages.


In [93]:
import pandas as pd
from datetime import datetime

In [ ]:
#Importing customer dataset
customer_dim = pd.read_excel("data/sales_database.xlsx", 
                             sheet_name="customer_dim")
customer_dim.head()

cust_id first_name last_name                     email       phone  \
0  10000001      Rahul     Gupta     rahul.gupta@gmail.com  8960722319   
1  10000002    Abhinav    Sharma  abhinav.sharma@gmail.com  8857852863   
2  10000003    Anubhav     Patel   anubhav.patel@gmail.com  8492146048   
3  10000004       Neha     Verma      neha.verma@gmail.com  8631861914   
4  10000005       John   Bernard    john.bernard@gmail.com  8215391844   

   primary_pincode  gender         dob joining_date  
0           110001    male  02-12-1977   19-02-2016  
1           400001    male  20-09-1995   01-10-2016  
2           560001    male  26-12-1986   16-08-2018  
3           600001  female  20-01-2002   28-07-2009  
4           500001    male         NaN   05-03-2017

In [ ]:
#Importing product dataset
product_dim = pd.read_excel("data/sales_database.xlsx", 
                            sheet_name="product_dim")
product_dim.head()

product_id     product_name brand  category  procurement_cost_per_unit  \
0       12345     HP XYZ Mouse    HP     mouse                        650   
1       12346          HP 241H    HP    laptop                      45000   
2       12347  HP 8GB Pendrive    HP  pendrive                        350   
3       12348   Dell ABC Mouse  Dell     mouse                        350   
4       12349       Dell AX420  Dell    laptop                      30000   

     mrp  
0   1500  
1  80000  
2    800  
3   1100  
4  75000

In [ ]:
#Importing delivery person dataset
delivery_person_dim = pd.read_excel("data/sales_database.xlsx",
                                    sheet_name="delivery_person_dim")
delivery_person_dim.head()

delivery_person_id             name joining_date  pincode
0             1000008  Indranoor Singh   08-03-2007   110001
1             1000002   Simon Williams   23-02-2007   400001
2             1000004    Anubhav Tyagi   16-11-2007   400001
3             1000006      Aviral Vats   16-11-2007   400001
4             1000009     Rocky Tiwari   16-11-2007   400001

In [ ]:
#Importing pincode dataset
pincode_dim = pd.read_excel("data/sales_database.xlsx", 
                            sheet_name="pincode_dim")
pincode_dim.head()

pincode       city       state
0   110001  New Delhi   New Delhi
1   400001     Mumbai  Maharastra
2   560001  Bangalore   Karnataka
3   600001    Chennai  Tamil Nadu
4   500001  Hyderabad   Telangana

In [ ]:
#Importing order dataset
order_dim = pd.read_excel("data/sales_database.xlsx", 
                          sheet_name="order_dim")
order_dim.head()

order_id order_type   cust_id  order_date delivery_date  tot_units  \
0  10000000001        buy  10000009  01-01-2020    03-01-2020          2   
1  10000000002        buy  10000004  01-01-2020    04-01-2020          9   
2  10000000003        buy  10000012  01-01-2020    02-01-2020          2   
3  10000000004        buy  10000012  01-01-2020    03-01-2020          8   
4  10000000005        buy  10000008  01-01-2020    06-01-2020          1   

   displayed_selling_price_per_unit  total_amount_paid  product_id  \
0                             73500             138180       12349   
1                             67500             577125       12349   
2                             60750             119070       12349   
3                               736               5824       12347   
4                             60750              57105       12349   

   delivery_person_id payment_type  delivery_pincode  
0             1000003         cash            500001  
1             1000005         cash            700001  
2             1000002   netbanking            400001  
3             1000010         cash            600001  
4             1000003        debit            500001

In [ ]:
%load_ext sql
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [80]:
%sql sqlite:///data/exercise_sql.db3

## Defina los esquemas de cada una de las tablas

In [107]:
%%sql sqlite://
DROP TABLE IF EXISTS customer_dim;
CREATE TABLE customer_dim(
    cust_id VARCHAR(50) NOT NULL UNIQUE,
    first_name VARCHAR(50) NOT NULL,  
    last_name VARCHAR(50) NOT NULL, 
    email VARCHAR(50) NOT NULL, 
    phone VARCHAR(50) NOT NULL, 
    primary_pincode VARCHAR(50)  NOT NULL,
    gender VARCHAR(6),
    dob DATETIME,
    joining_date DATETIME, 
    PRIMARY KEY (cust_id)
);

Done.
Done.


[]

## Poble cada una de las tablas

In [108]:
for row in range(len(customer_dim)):
    cust_id = str(customer_dim.iloc[row,0])
    first_name = customer_dim.iloc[row,1]
    last_name = customer_dim.iloc[row,2]
    email = customer_dim.iloc[row,3]
    phone = str(customer_dim.iloc[row,4])
    primary_pincode = str(customer_dim.iloc[row,5])
    gender = customer_dim.iloc[row,6]
    
    try:
        dob_item = datetime.strptime(customer_dim.iloc[row,7], '%d-%m-%Y')
    except:
        dob_item = None
        
    try:
        joining_date = datetime.strptime(customer_dim.iloc[row,8], '%d-%m-%Y')
    except:
        joining_date = None
    
    
    %sql INSERT INTO customer_dim VALUES (:cust_id, :first_name, :last_name, :email, :phone, :primary_pincode, :gender, :dob_item, :joining_date); 

 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.
 * sqlite:///data/exercise_sql.db3
1 rows affected.


In [109]:
%sql SELECT * FROM customer_dim

 * sqlite:///data/exercise_sql.db3
Done.


[('10000001', 'Rahul', 'Gupta', 'rahul.gupta@gmail.com', '8960722319', '110001', 'male', '1977-12-02 00:00:00', '2016-02-19 00:00:00'),
 ('10000002', 'Abhinav', 'Sharma', 'abhinav.sharma@gmail.com', '8857852863', '400001', 'male', '1995-09-20 00:00:00', '2016-10-01 00:00:00'),
 ('10000003', 'Anubhav', 'Patel', 'anubhav.patel@gmail.com', '8492146048', '560001', 'male', '1986-12-26 00:00:00', '2018-08-16 00:00:00'),
 ('10000004', 'Neha', 'Verma', 'neha.verma@gmail.com', '8631861914', '600001', 'female', '2002-01-20 00:00:00', '2009-07-28 00:00:00'),
 ('10000005', 'John', 'Bernard', 'john.bernard@gmail.com', '8215391844', '500001', 'male', None, '2017-03-05 00:00:00'),
 ('10000006', 'Muhammad', 'Ali', 'muhammad.ali@gmail.com', '8317878411', '700001', 'male', '1981-09-29 00:00:00', '2019-03-13 00:00:00'),
 ('10000007', 'Ahmed', 'Khan', 'ahmed.khan@gmail.com', '8279227825', '110001', 'male', '1999-01-14 00:00:00', '2018-07-18 00:00:00'),
 ('10000008', 'Paras', 'Rana', 'paras.rana@gmail.com', '8134672373', '400001', 'male', '1985-09-20 00:00:00', '2015-02-23 00:00:00'),
 ('10000009', 'Priyanka', 'Reddy', 'priyanka.reddy@yahoo.com', '8156273277', '560001', 'female', '1992-08-26 00:00:00', '2010-11-24 00:00:00'),
 ('10000010', 'Amit', 'Vikram', 'amit.vikram@yahoo.com', '8204474821', '600001', 'male', '1996-04-03 00:00:00', '2013-08-25 00:00:00'),
 ('10000011', 'Sandhya', 'Anwar', 'sandhya.anwar@outlook.com', '8700166174', '500001', 'female', '2002-04-11 00:00:00', '2009-03-08 00:00:00'),
 ('10000012', 'Robert', 'Smith', 'robert.smith@outlook.com', '8933123477', '700001', 'male', None, '2017-11-16 00:00:00'),
 ('10000013', 'Christina', 'Smith', 'christina.smith@outlook.com', '8861486813', '700001', 'female', '1980-04-16 00:00:00', '2009-11-30 00:00:00'),
 ('10000014', 'Amanpreet', 'Singh', 'amanpreet.singh@yahoo.com', '8153506539', '110001', 'male', '1980-03-10 00:00:00', '2008-08-29 00:00:00')]

## Preguntas sobre SQL básico

### Q0.¿Cuántos clientes son hombres y cuántos son mujeres?

In [110]:
%%sql
SELECT 
    c.gender,
    COUNT(c.cust_id) AS cust_cnt
FROM customer_dim AS c
GROUP BY c.gender


 * sqlite:///data/exercise_sql.db3
Done.


[('female', 4), ('male', 10)]

### Q1. ¿Cuántos clientes no disponen de información sobre la fecha de nacimiento?

### Q2. ¿Cuántos clientes hay en cada combinación de código postal y género?

### Q3. Muestre el nombre del producto y el MRP para los productos que tienen más de 50000 MRP

### Q5. Para cada código Pin, imprima el recuento de órdenes, la suma del importe total pagado, el importe medio pagado, el importe máximo pagado, el importe mínimo pagado para las transacciones que se pagaron en "efectivo". Tome sólo los tipos de órdenes de "compra".

### Q6. Para cada delivery_person_id, imprima el recuento de pedidos y el importe total pagado para product_id = 12350 o 12348 y unidades totales > 8. Ordene la salida por el importe total pagado en orden descendente. Tome sólo los tipos de órdenes de "compra".

### Q7. Imprima los nombres completos (nombre más apellido) de los clientes que tienen correo electrónico en "gmail.com"(investigar comando ***LIKE*** https://www.sqlitetutorial.net/sqlite-like/)

### Q8. ¿Cuántas órdenes tenían #unidades entre 1-3, 4-6 y 7+? Tome sólo los tipos de órdenes de "compra".

### Q9. ¿Qué código postal tiene un importe medio pagado superior a 150.000? Tome sólo los tipos de órdenes de "compra".

### Q10. Cree las siguientes columnas a partir de los datos de order_dim -

- Día del pedido (order_day)
- Mes del pedido (order_month)
- Año del pedido (order_year)

### Q11. ¿Cuántos pedidos totales hubo en cada mes y cuántos de ellos fueron devueltos? Añade también una columna para la tasa de devolución.

tasa de devolución = (100.0 * total de órdenes de devolución) / total de órdenes de compra

Sugerencia: Tendrá que combinar SUM() con CASE WHEN

## Preguntas sobre SQL Joins

### Q12. ¿Cuántas unidades se han vendido de cada marca? Obtén también el total de unidades devueltas de cada marca.

### Q13. ¿Cuántos clientes y repartidores distintos hay en cada estado?

### Q14. Para cada cliente, imprima cuántas unidades totales se pidieron, cuántas unidades se pidieron desde su primary_pincode y cuántas se pidieron no desde el primary_pincode. Calcule también el porcentaje de unidades totales que se pidieron desde primary_pincode (recuerde multiplicar el numerador por 100,0). Ordene por la columna de porcentaje en orden descendente.

### Q15. Para cada nombre de producto, imprima la suma del número de unidades, el importe total pagado, el precio de venta total mostrado, el precio mínimo total de estas unidades y, finalmente, el descuento neto del precio de venta (es decir, 100,0 - 100,0 importe total pagado / precio de venta total mostrado) Y el descuento neto del precio mínimo (es decir, 100,0 - 100,0 importe total pagado / precio mínimo total)


### Q16. Para cada order_id (excluyendo las devoluciones), obtenga el nombre del producto y calcule el porcentaje de descuento del precio de venta. Ordene por el mayor descuento e imprima sólo las filas en las que el porcentaje de descuento sea superior al 10,10%.

### Q17. Utilizando el coste de aprovisionamiento por unidad en product_dim, encuentra qué categoría de producto ha obtenido el mayor beneficio tanto en cantidad absoluta como en porcentaje

Beneficio absoluto = Cantidad total vendida - Coste total de aprovisionamiento

Beneficio porcentual = 100.0 * Cantidad total vendida / Coste total de aprovisionamiento - 100.0

### Q18. Para cada repartidor (utilice su nombre), imprima el número total de pedidos (excluyendo las devoluciones) por mes en columnas separadas, es decir, debe haber una fila para cada delivery_person_id y 12 columnas para cada mes del año

### Q19. For each gender - male and female - find the absolute and percentage profit (like in Q16) by product name.

### Q20. Generally the more numbers of units you buy, the more discount seller will give you. For 'Dell AX420' is there a relationship between number of units ordered and average discount from selling price? Take only 'buy' order types